### DISCUSSION TITLE
[Feature Request] Logging

### DISCUSSION THREAD

**Discussion by BillyCroan** [memtest86plus/memtest86plus#360] [Enhancement]
The ability to log test progress and errors would bring great value to this project and make it usable on headless enterprise systems.

What would you think about logging via syslog?  (single udp packet per event)
There's a linux kernel module called netconsole that does this with a simplified network stack.  boot arguments would convey which interface, mac address, port number to use, and no bidirectional communication is required

Or logging via serial port may also help.  Many systems today implement serial redirection over ssh or ipmi sol.  boot arguments for serial logging would specify which serial port to use and possibly the baud rate.

**Comment by x86fr** [comment]
Accessing network requires building a NIC drivers from scratch. Actually writing several hundreds of drivers for every possible NICs. Clearly not possible. The "easier" solution will be to write a USB mass-storage driver with a FAT32 filesystem driver.

Logging via serial port is already implemented.

**Comment by debrouxl** [comment]
* memtest86+ supports serial output in a number of conditions, governed by the `console=` command-line parameter, per the README. Is your target platform unsupported by current memtest86+ versions ?

* implementing even reduced networking features for memtest86+ is a much more involved process. Modern UEFI BIOS implementations provide some networking support, but memtest86+ intentionally exits UEFI boot services, in order to be able to test memory which would otherwise be locked by the UEFI BIOS, up to hundreds of megabytes on some platforms. memtest86+ also needs to work on so-called legacy BIOS anyway. Providing the ability to log data through the network (or onto FAT32, ext*, or other filesystems stored on USB Mass Storage Devices, individual internal IDE / SCSI / SATA / SAS / NVMe / etc. drives, RAID controllers for groups of drives) requires implementing everything into memtest86+ - network drivers, IP stack, disk controller drivers, filesystem drivers, etc.

Sure, some existing RTOS provide support for disk+filesystem access and networking, for a subset of drive controller types and NICs. I've never checked whether any of these RTOS actually supports the special-purpose periodic dynamic relocation of its own self at runtime that memtest86+ implements, in order to maximize the amount of memory tested by the program, but there's a good chance that not all of them support dynamic relocation... assuming that any of them supports it, even.

**Reply by changpenghui616** [reply]
I use the last version to test，run command in shell，command is ‘memtest.efi console=ttyS’, but no out put at serial port。im use memtest.efi console=ttyS0、memtest.efi console=ttyS1、memtest.efi console=ttyS2、memtest.efi console=ttyS3 too，still no output

**Reply by debrouxl** [reply]
Does the patch in #419 change the behaviour for you ?

**Reply by changpenghui616** [reply]
No,it still cant output serial log when use this patch

**Comment by tormodvolden** [comment]
Since many computers nowadays don't have serial ports (or any built-in emulation) it would be nice to add support for common USB-serial/UART adapters like those based on cp2102, pl2303, ftdi etc. But something that would be much simpler on the memtest86+ side is making a custom firmware for a USB dongle (based on STM32/libopencm3, RP2040 or Arduino Micro or similar) that implements a USB keyboard but also adds a custom USB request to "print" a character that the dongle can forward over serial/wifi/etc to another computer. This should be fairly easy to hook into the existing USB keyboard code.

**Comment by tormodvolden** [comment]
I am interested in what the maintainers think about this. It would be most "proper" to add support for common USB-serial adapters to the memtest86+ code. However there are so many different adapters versions and variants, so it opens up for a support burden "I bought this USB serial adapter on internet and it doesn't work with memtest86+" ... It is also quite difficult to buy an adapter with a given chipset and revision, if we decide to support only a few of them.

Anyway, a proof of concept is shown in PR #387.

OTOH a new "memtest gadget" device would mean minimal changes and support in the memtest86+ code and leave most issues to the gadget firmware developers instead.

**Comment by debrouxl** [comment]
First of all, thanks for spending time working on that proof of concept for an interesting take on the USB serial logging capability :)

On principle, I'm not against the idea, as the logging functionality is clearly useful to some users (when equipped with the appropriate hardware), and your PoC makes it clear that implementing output of screen data to one model of USB serial adapter doesn't add significant bloat to the memtest86+ code base - unlike one NIC driver, certainly.

Support for a reasonable set of CP210x + PL2303 + FTDI232* + CH34x controller hardware revisions, which seem to represent a sizable part of the USB serial adapter market, would make the feature useful to more users, but might also bloat the code base somewhat more, if lots of hardware quirks - or simply hundreds of USB VIDs and PIDs - are to be taken into account ?

**Comment by x86fr** [comment]
Nice PoC and thanks for your work. Here are some thoughts in random form:

Serial console output are primarily used on servers with virtual COM port (now supported) and almost all modern motherboards still have a physical COM header (although without the DE9 connector attached). This said, just having to plug a USB/RS232 into a USB port to get logging would be interesting if we can at least support CP210x/FTDI232/CH34x and EHCI/xHCI. 

However, if anybody is willing to spend some dev time on logging feature, maybe the best way to go is to work on USB Flash drive support for logging. There is already many open-source implementation of the standardized mass storage device specifications (same for FAT32 filesystem).

**Comment by tormodvolden** [comment]
Something I wanted to mention is that for logging we would need, independent of where the output will go, a new logging mode. The serial console mode mirrors the original screen using terminal emulation codes (VT100/ANSI?). This is unnecessarily verbose and not suitable for storage, or automated processing. (In the meantime #389 came as an example.)
